### PR TITLE
Integration of SCT into fsleyes UI

### PR DESCRIPTION
**Pull request by perone**
This is the PoC for the SCT integration into the fsleyes. I've implemented the integration for `sct_deepseg_gm` and also for the `sct_deepseg_sc`. Screenshots below:

# Comments
The integration itself is done, which I found to be the tricky part (to dynamically change the UI, given that they didn't provided any examples of how to do it), however, there are of course many improvements that can be made, such as:

- UX can be improved (button icons, positioning, help, etc);
- Add an activity bar (i.e. progress bar for activity) to show that the process is running, given that it may take a while for the segmentation to finish;
- Feedback when something goes wrong (adding a new window containing the command output log is an alternative that many UIs integrating command-line tools use);
- Integrate more tools;
- Redesign the code to have a window pattern that can be used by many different tools (just extend it and reuse);
- Add more options (keep temporary files, where the output file will be saved, etc).
- Integrating the SCT tools using the SCT APIs (as the refactoring happens) instead of calling the external command-line;

Anyway, I consider this a good starting point that is functional. It seems that `fsleyes` is robust enough and already contains a lot of widgets and useful tools to handle the volume data. The best part is that it is written in Python, giving us a lot of flexibility to **control the entire UI components**.

# How to use it ?
Just enter in `fsleyes`, click in `Run script...` and select the `sct_plugin.py` file.

# Screenshots
### 1) Message when SCT isn't found in the system:
image

### 2) Interface for the Gray Matter Segmentation (`sct_deepseg_gm`):
image

### 3) Interface for the Spinal Cord Segmentation (`sct_deepseg_sc`):
image

### 4) Example after clicking in in both segmentations:
image

### CODE REVIEW

**Comment by jcohenadad** [comment]
another comment apart from the logo: instead of "SCT / Segment Gray Matter", how about calling it "sct_deepseg_gm" instead? So that people know what function is called. Same for sct_deepseg_sc.

Another thing: is it possible to display the "usage description" within the fsleyes plugin area?

**Comment by perone** [comment]
I think it would be better if we could have this logo with transparency. Otherwise, it will have this white background like the image below:

image

I also changed the tab titles. About the "usage description", we can add a text in the right or left (on the right side of the logo) with a small help like:
```
This is the Spinal Cord (SC) segmentation tool.
Instructions:
1. Select the contrast;
2. Click in the button X;
```

What do you think ?

**Comment by jcohenadad** [comment]
@perone 

Having a usage description that says "1. Select the contrast; 2. Click in the button X;" is not very relevant (it is pretty obvious where to click). I was thinking more of an extended usage description, that includes the type of algorithm it is using, as well as a reference to cite. E.g. for deepseg_gm:
~~~
Spinal Cord Gray Matter (GM) Segmentation using deep dilated convolutions. Reference: CS Perone, E 
Calabrese, J Cohen-Adad. Spinal cord gray matter segmentation using deep dilated convolutions 
(2017). arxiv.org/abs/1710.01269
~~~
To avoid duplication, you can probably access that field directly from the parser.

Here is a logo with alpha layer:
logo_sct-alpha
You can try to make the logo much smaller, and have the description below it.

**Comment by perone** [comment]
Make sense, I'll experiment with different sizes and will add also add a new label with the description, thanks.

**Comment by perone** [comment]
Ok so I refactored the code with the following changes:

- Create a class called `SCTPanel` that tools can extend and keep the same design;
- Calls to the SCT tools are now on another separate thread, to avoid blocking the UI;
- The UI will now show a modal dialog when the algorithm is running;
- The logo was updated in the repo;

## Screenshots

image

image

image

**Comment by jcohenadad** [comment]
Much better. Could you please also add the following citation below the SCT logo:
~~~
De Leener B et al. SCT: Spinal Cord Toolbox, an open-source software for processing spinal cord MRI data. Neuroimage. 2017 Jan 15;145(Pt A):24-43.
~~~
thx!

File: contrib/fsl_integration/sct_plugin.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    notebook.AddPage(panel_gm, "sct_deepseg_gm", True)
+    notebook.AddPage(panel_sc, "sct_deepseg_sc", False)
+
+    aui_manager.AddPane(notebook, 
+                        aui.AuiPaneInfo().Name("notebook_content").
+                        CenterPane().PaneBorder(False)) 
+    aui_manager.Update()
+
+
+run_main()
```
Comment (peristeri) [inline contrib/fsl_integration/sct_plugin.py]: Replace with: 
```
if __name__ == "__main__":
    run_main()
```
it would make it the file into a module.

Comment (perone) [reply contrib/fsl_integration/sct_plugin.py]: Note that this script is being executed by fsleyes, so the `__name__` won't be set to `__main__`. If I add this checking, the `run_main()` will not execute.

PS: `wx` isn't installed on SCT, it is part of the fsleyes.

**Comment by perone** [comment]
Added SCT citation:

image

File: contrib/fsl_integration/sct_plugin.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+import wx
+import wx.lib.agw.aui as aui
+import wx.html as html
+
+
+aui_manager = frame.getAuiManager()
+
+
+def sct_call(command):
```
Comment (peristeri) [inline contrib/fsl_integration/sct_plugin.py]: This function is a copy of line 43. Is this function needed?

Comment (perone) [reply contrib/fsl_integration/sct_plugin.py]: Good catch, I removed it. Thanks !

**Comment by jcohenadad** [comment]
@perone i'm trying to fix a couple of things (e.g., file created within SCT install folder). How can I visualize the process running for debugging purpose?

**Comment by perone** [comment]
You can add a `print()` in these variables:

    stdout, stderr = p.communicate()

like:

    print(stdout)
    print(stderr)

These are the outputs of the function execution.

**Comment by jcohenadad** [comment]
i suggest we leave the print, this is very useful!

**Comment by pauldmccarthy** [comment]
I just found this via a random google search - it looks super cool, and I'm stoked that you're hacking FSLeyes!

I have been thinking for a while about adding a simple "plugin" architecture, which will allow you to write your code in a slightly more formal manner - I've started writing some notes on this here: https://git.fmrib.ox.ac.uk/fsl/fsleyes/fsleyes/issues/110

Feel free to send me an email if you have any thoughts on this (unfortunately our internal gitlab server doesn't have the option for somebody without an account to add/comment on issues). I'll hopefully be able to start working on this in the next month or two. 

And you've probably already found it, but I have the FSLeyes API documentation hosted here: https://users.fmrib.ox.ac.uk/~paulmc/fsleyes/apidoc/latest/fsleyes.html

**Comment by perone** [comment]
Hi @pauldmccarthy, great thanks for your message and links for the API ! It's very cool that you guys are planning to implement a plugin architecture, that would be awesome for many people willing to integrate their tools in fsleyes (like us !). I'll add some comments below that you might find interesting and that can be useful both for developers and users:

- It might happen that plugins would require extra Python module dependencies (or images, data, etc), and since fsleyes has its internal Python interpreter bundled, it would be interesting to implement the plugins as Python modules that would be installed inside this internal bundled Python distribution. This is very nice because you can use the Python native entry-points to search for plugins once a package is installed (see [here](https://setuptools.readthedocs.io/en/latest/pkg_resources.html#entry-points) and [here](https://stackoverflow.com/a/9615473/179372) for instance). Many Python packages uses this entry-point to manage the plugin architecture. Developers can create their plugins with names such as `fsleyes-plugin-[plugin name]` for instance, and then register it in official Python PyPI repository with a `setup.py` defining the entry-point that fsleyes will use. That way developers can define their requirements and installation in their own plugin `setup.py` file.

- To be easy for end-users to install a new plugin, you can specify a json metadata file that can be created by plugin developers, where it mentions the package name, description, version, link to the package, etc. This way users can just click on an `"Install Plugin..."` action in fsleyes that will ask for the URL of this metadata file and you can then fetch it and install this package on the internal bundled Python distribution using pip programmatically (like [this](https://gist.github.com/rwilcox/755524) for instance). Another alternative, that takes more work is to create a screen that fetches the available (and approved) extensions from your central repository and shows a list (like [this](http://codevista.net/wp-content/uploads/2015/08/Library-Manager.png) for instance, taken from Arduino IDE), where users can simply click to install or remove the plugin. This last option is very good for the users because thay can see everything available, install/remove, etc, and you can also have control of what users see because this list is retrieved from your repository;

Anyway, these were some ideas but of course it depends on the amount of efforts you guys are planning to spend on it. I really liked the proposal that you made to have an option to install the .py file and the architecture described in [fsleyes - Issue #110](https://git.fmrib.ox.ac.uk/fsl/fsleyes/fsleyes/issues/110), it would be **really** amazing and helpful to have something like that even without the suggestions I made above. Thanks again for entering in contact and let us know about it !

**Comment by pauldmccarthy** [comment]
Hi @perone, thanks for the thoughts - I do like the idea of using PyPI, and installing plugins into the environment. I'll have to have a play around to see if this can work in environments where the user doesn't have admin rights (e.g. multi-user servers), and for PyPI installs of FSLeyes (as opposed to the standalone version, as is bundled with FSL). Perhaps a hybrid approach will work. I'll get back to you when I get a chance to work on this!

**Comment by perone** [comment]
@pauldmccarthy makes sense, thanks again for getting in touch !

**Comment by pauldmccarthy** [comment]
Hey @perone, the beginnings of a FSLeyes plugin architecture are included in the latest version of FSLeyes (0.26.1)! 

 - Docs: https://users.fmrib.ox.ac.uk/~paulmc/fsleyes/userdoc/latest/plugins.html
 - Code: https://git.fmrib.ox.ac.uk/fsl/fsleyes/fsleyes/blob/master/fsleyes/plugins.py
 - Minimal ``setup.py``-based example: https://git.fmrib.ox.ac.uk/fsl/fsleyes/fsleyes/tree/master/tests/testdata/fsleyes_plugin_example

It is *very* simple at this stage. All views, controls, and tools that are provided by the loaded plugins essentially share a global namespace (with names defined via ``setup.py`` entry points, or just class names for ``.py`` based plugins). So if two plugins provide a control with the same name, only one of them will appear in the FSLeyes interface.

There is also no facility for querying available plugins on PyPi - I am hoping that the [Warehouse API](https://warehouse.pypa.io/api-reference/json/) will be extended in the future to allow packages to be searched by name. The only option right now (as far as I could determine) is to download a list of *all* packages on PyPi, and then grep them after the request.

If you have any thoughts on how the framework could be improved to make your life easier, feel free to open an issue over at [pauldmccarthy/fsleyes](https://github.com/pauldmccarthy/fsleyes).

**Comment by jcohenadad** [comment]
Fantastic news! Thank you for the heads-up @pauldmccarthy. We will definitely give it a try.

**Comment by perone** [comment]
Thanks @pauldmccarthy, these are great news, we'll certainly take a look at it, thanks a lot for providing this modularity, it's very exciting for many projects.

**Pull request merged**